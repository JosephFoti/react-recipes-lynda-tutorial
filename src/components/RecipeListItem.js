import React from 'react';
import PropTypes from 'prop-types';

const RecipeListItem = ({ recipe, onClick, onFavorited, favorited }) => (
  <li
    className="py2 border-bottom border-bottom-dashed border-width pointer"
    key={recipe.id}
    onClick={() => onClick(recipe.id)}
  >
    <span
      className="mr1"
      onClick={e => {
        e.stopPropagation();
        onFavorited(recipe.id);
      }}
      role="img"
      aria-label="favorite"
    >
      {favorited ? '✅' : '◻️'}
    </span>
    <span>{recipe.name}</span>
    <span>{recipe.category}</span>
  </li>
);

RecipeListItem.propTypes = {
  recipe: PropTypes.object,
  favorited: PropTypes.bool,
  onClick: PropTypes.func,
  onFavorited: PropTypes.func,
};

RecipeListItem.defaultProps = {
  recipe: {},
};

export default RecipeListItem;
