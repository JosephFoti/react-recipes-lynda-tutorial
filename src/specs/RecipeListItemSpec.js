import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import RecipeListItem from '../components/RecipeListItem';

const testRecipe = {
  id: 1,
  name: 'Test Recipe',
  category: 'Test category',
};

/*

Testing with Jest
config syntax -> jest.config.js

*/

describe('<RecipeListItem />', () => {
  test('Should not break if no recipe passed', () => {
    const component = renderer.create(<RecipeListItem />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
  test('Should Correctly Render', () => {
    const component = renderer.create(<RecipeListItem recipe={testRecipe} />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
  test('Should render favorited state', () => {
    const component = renderer.create(
      <RecipeListItem recipe={testRecipe} favorited />,
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
  test('Should render onClick when clicked', () => {
    const onClick = jest.fn();

    const component = mount(
      <RecipeListItem recipe={testRecipe} onClick={onClick} />,
    );
    component.simulate('click');

    expect(onClick.mock.calls.length).toBe(1);
  });
  test('Should render onFavorited when favorited', () => {
    const onFavorited = jest.fn();

    const component = mount(
      <RecipeListItem recipe={testRecipe} onFavorited={onFavorited} />,
    );
    component
      .find('span')
      .first()
      .simulate('click');

    expect(onFavorited.mock.calls.length).toBe(1);
  });
  test('Should not call onClikc when onFavorited is called', () => {
    const onClick = jest.fn();
    const onFavorited = jest.fn();

    const component = mount(
      <RecipeListItem
        recipe={testRecipe}
        onClick={onClick}
        onFavorited={onFavorited}
      />,
    );
    component
      .find('span')
      .first()
      .simulate('click');

    expect(onClick.mock.calls.length).toBe(0);
    expect(onFavorited.mock.calls.length).toBe(1);
  });
});
