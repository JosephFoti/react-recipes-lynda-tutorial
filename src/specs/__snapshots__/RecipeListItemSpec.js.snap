// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecipeListItem /> Should Correctly Render 1`] = `
<li
  className="py2 border-bottom border-bottom-dashed border-width pointer"
  onClick={[Function]}
>
  <span
    aria-label="favorite"
    className="mr1"
    onClick={[Function]}
    role="img"
  >
    ◻️
  </span>
  <span>
    Test Recipe
  </span>
  <span>
    Test category
  </span>
</li>
`;

exports[`<RecipeListItem /> Should not break if no recipe passed 1`] = `
<li
  className="py2 border-bottom border-bottom-dashed border-width pointer"
  onClick={[Function]}
>
  <span
    aria-label="favorite"
    className="mr1"
    onClick={[Function]}
    role="img"
  >
    ◻️
  </span>
  <span />
  <span />
</li>
`;

exports[`<RecipeListItem /> Should render favorited state 1`] = `
<li
  className="py2 border-bottom border-bottom-dashed border-width pointer"
  onClick={[Function]}
>
  <span
    aria-label="favorite"
    className="mr1"
    onClick={[Function]}
    role="img"
  >
    ✅
  </span>
  <span>
    Test Recipe
  </span>
  <span>
    Test category
  </span>
</li>
`;

exports[`<RecipeListItem /> Should render onFavorited when favorited 1`] = `
<li
  className="py2 border-bottom border-bottom-dashed border-width pointer"
  onClick={[Function]}
>
  <span
    aria-label="favorite"
    className="mr1"
    onClick={[Function]}
    role="img"
  >
    ✅
  </span>
  <span>
    Test Recipe
  </span>
  <span>
    Test category
  </span>
</li>
`;
